### ISSUE TITLE
Issue with SyntaxWalker and Malformed XML comments

### ISSUE THREAD

**Issue by paul1956** [dotnet/roslyn#26973]
**Version Used**: 
2.8 Libraries,  C# and VB.Net, Visual Studio 15.7.1
**Steps to Reproduce**:
```
    ''' <summary> ' XML Element without End Tag
    '''   Some XML text that is part of summary
    '''' This is a comment and not part of XML Summary
    ''' </summary> ' This is XML Text and not a close Summary

    /// <summary> /* XML Element without End Tag */
    ///   Some XML text that is part of summary
    //// This is a comment and not part of XML Summary
    /// </summary> /*  This is XML Text and not a close Summary */
```
1. In C#, VB and Visual Studio the above XML document comment is allowed but the SyntaxWalker returns unexpected results and you can't create the above with SyntaxFactory.XmlElement(StartTag, Content, EndTag) if the end tag is empty. 
2. The Visitor for the XmlElement node contains the Summary element, Content element with an empty EndTag

**Expected Behavior**:
The compilers and Visual Studio should complain about Malformed Document comments or the SyntaxWalker should treat the embedded comments as part of the content
**Actual Behavior**:
No errors in Compilers or Visual Studio but SyntaxWalker passes in malformed XML

**Comment by CyrusNajmabadi** [comment]
>  but the SyntaxWalker returns unexpected results

Can you explain what this means?

**Comment by CyrusNajmabadi** [comment]
> No errors in Compilers or Visual Studio but SyntaxWalker passes in malformed XML

Can you explain what you mean by that?  I get:

![image](https://user-images.githubusercontent.com/4564579/40256846-c4bd5b88-5aa0-11e8-909e-63022571aba2.png)

Have you actually asked the compiler to generate the xml doc file?

**Comment by paul1956** [comment]
@CyrusNajmabadi I tried to simplify the issue and guess I failed. In many files throughout Roslyn (for example roslyn-master\src\Compilers\CSharp\Test\Syntax\Resources.Designer.cs starting at line 76) the space after the CS.SyntaxKind.DocumentationCommentExteriorTrivia is missing. Those lines are treated as comments and not part of the XML Document Comment. I am using the XMLVisitors to walk the comments and the EndTag is empty.

The file above has no errors and no warnings, but the color coding show what I described in the original issue.

```
       Public Overrides Function VisitXmlElement(node As CSS.XmlElementSyntax) As VisualBasicSyntaxNode
            Dim StartTag As Syntax.XmlElementStartTagSyntax = CType(node.StartTag.Accept(Me).WithConvertedTriviaFrom(node.StartTag), Syntax.XmlElementStartTagSyntax)
            Dim Content As New SyntaxList(Of Syntax.XmlNodeSyntax)
            Dim EndTag As Syntax.XmlElementEndTagSyntax
            EndTag = SyntaxFactory.XmlElementEndTag(CType(StartTag.Name, Syntax.XmlNameSyntax))
           ' skip getting content
           Return SyntaxFactory.XmlElement(StartTag, Content, EndTag)
       End Function
```

**Comment by CyrusNajmabadi** [comment]
Could you make a small repro where you see this problem, along with the code you are running that demonstrates the issue when using hte API?

**Comment by paul1956** [comment]
I don't have a simple Repo (but if necessary I will create one) the code I am running is above, but you can see the behavior without any code using Visual Studio. When you open the file listed above you will see the /// in gray when the XML is correctly formed and Green when it is incorrect and you will see VS report no warnings. In the example below I have added spaces after the /// in the first document comment but not the second and all the /// turn gray and the XML parses/visits correctly.

![image](https://user-images.githubusercontent.com/12789369/40276008-01f9e6e2-5bb4-11e8-8183-2575761c5232.png)

The SyntaxVisualize also shows the Warnings and highlights the issues in Pink

![image](https://user-images.githubusercontent.com/12789369/40276055-0fc82f62-5bb5-11e8-8acc-315f3249a464.png)

Please correct me if I am wrong but I would expect Document Comments in Roslyn to be correctly formed or for Visual Studio and the compiler to complain. It appears that comments are allowed in the middle of an XML Document Comment and if that is the case then the VIsitElement should parse it correctly.

**Comment by paul1956** [comment]
One note the picture I originally posted 9 hours ago with gray slashes was not from Visual Studio is was from my application. The pictures directly above are from Visual Studio.

**Comment by paul1956** [comment]
@CyrusNajmabadi could you point to any Test Code in Roslyn that tests VisitXmlElement, I can't find any, then it would be easy to change the test to reproduce the issue. The bug I am trying to report is either in the code that calls VisitorXML.. with Malformed XML or in the definition of where comments are allowed in an XMLDocumentComment.

**Comment by paul1956** [comment]
ping After hitting this again. It seems that with the C# XML below
```C#
        /// <summary>
        /// Verify result.
        /// child::para[1] <= child::para[2]
        /// </summary>

        /// <summary>
        /// Verify result.
        /// child::para[1] < child::para[2]
        /// </summary>
```
The 3rd line in both examples should be a single XML element but it is actually split into multiple XML items depending on the character following the < including XMLEmptyToken and Trailing Trivia.  The second example split into an XMLElement and and XMLEmptyElement. There are a lot of other example but generally from the perspective of the Syntax Walker "<" within an XML Element creates a New XML element.

**Comment by CyrusNajmabadi** [comment]
@paul1956 you need to use `&lt;` for a `<` sign in xml.

**Comment by paul1956** [comment]
@CyrusNajmabadi this isn't my code, it is in the Source of CoreFx the compilers have no issue compiling it and VS shows no "error" just the wrong coloring, only the Roslyn SyntaxWalker crashes trying to parse it.

**Comment by CyrusNajmabadi** [comment]
> only the Roslyn SyntaxWalker crashes trying to parse it.

What crash do you get?

**Comment by CyrusNajmabadi** [comment]
> the compilers have no issue compiling it

Yes they do.  Check it out in VS:

> and VS shows no "error"

Yes it does.  Check it out:
![image](https://user-images.githubusercontent.com/4564579/52113548-75b01a00-25be-11e9-9b29-e1628e9ae87b.png)

**Comment by CyrusNajmabadi** [comment]
> There are several places where malformed XML Document comments are accepted by the compiler and VS but not Roslyn.

Roslyn *is* the compiler, and it is the Language Service for C#/Vb in VS.  So your statement doesn't make sense.

Is it possible you have this unchecked:

![image](https://user-images.githubusercontent.com/4564579/52113615-adb75d00-25be-11e9-9ac9-e823274bb69a.png)

**Comment by CyrusNajmabadi** [comment]
> it is in the Source of CoreFx 

That doesn't mean it's correct.  It's likely that CoreFx simply has an incorrect doc-comment.  If you process incorrect code with Roslyn, you will likely see unexpected results in the syntax nodes and whatnot.  It's the same as if i were to write:

```c#
now is the time for all good men
```

The C# parser will parse this into *something*, but it's not necessarily going to be anything good, or anything you might expect to be able to easily process.

**Comment by paul1956** [comment]
VS 2017 shows no warning or suggestions.
![image](https://user-images.githubusercontent.com/12789369/52117102-c1b38c80-25c7-11e9-87ff-4c8eeda9b820.png)
Using Roslyn to parse a program that VS 2017 does not show any errors for. It looks like whatever version of VS you are running has theses issues fixed. I am only trying to parse code that VS thinks has no errors. After loading the project that contains the file and checking the XML Documentation File box I see no difference. Thanks I think I learned something. My guess when the next version of VB comes out that shows the error, the code will get fixed. For now I will ignore the errors. I guess this issue can be marked and fix (in some future version of VS). For now I will convert it to &lt;., it also explains why I can't create a VB XML node with "<". Thanks for the prompt response.

**Comment by sharwell** [comment]
@paul1956 When you don't have the option enabled to create an XML documentation comment file, Roslyn treats `///` comments the same as it treats `//` or `////` comments. The syntax highlight still tries to highlight them, but from the compiler perspective *there are no documentation comments* under that configuration. This is a constant source of user problems for StyleCop Analyzers, so we created [SA0001](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA0001.md) to point users in the right direction.

When you enable documentation comment output, the warnings about invalid files show up where you expect.

**Comment by CyrusNajmabadi** [comment]
> I guess this issue can be marked and fix 

Can you explain what fix you are looking for?  Right now, everything seems by design

1. You do not have the vs option to generate did comments. So warnings about the doc comment are *suppressed* from the ui, but are still there in the tree.
2. If you parse the tree, the nodes will have problems.  

This is all by design and I don't believe anything will be fixed here.

**Comment by paul1956** [comment]
From the comments from @sharwell it looks like I need **an option** so that the Roslyn Syntax Walker works the same way as VS and treats //, /// and /// with or without a trailing space all as comments and not as malformed XML Comments. I don't want to fix existing C# code, I want to translate it to VB and have the VB compiler compile it. If It happens to have exactly 3 slashes followed by a space and no <, the VB translation will treat it as an XML Document comment. So the question is how do I configure the Roslyn Syntax Walker to simulate the effect of the check box being unchecked.

1. There are no diagnostic is the tree, all the programs I am testing compile without error or warnings assume the box is not checked, the syntax Walker attempts to parse the XML and creates a lot of EmptyXMLTokens with comments but no errors.

**Comment by paul1956** [comment]
Thanks @sharwell. I found the issue, what I need is already configurable, I need to set DocumentationMode to None instead of Parse.
